File: TRViS.UITests/Tests/IsInfoRowTransitionTests.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Appium.iOS;
using OpenQA.Selenium.Appium.Mac;
using TRViS.UITests.Pages;

namespace TRViS.UITests.Tests;

/// <summary>
/// Regression test for "non-InfoRow components (station name, drive time, etc.)
/// remain visible after a WebSocket soft-update changes IsInfoRow from false to true
/// on a timetable row."
///
/// Uses the UI_TEST seam <c>DTAC.TestSeedIsInfoRowTransitionButton</c> which:
///   1. Takes the first non-InfoRow in the currently selected train,
///   2. Clones the TrainData with that row's IsInfoRow flipped to true, and
///   3. Re-sets AppViewModel.SelectedTrainData — triggering the same soft-update
///      code path as a WebSocket edit (same train ID → ApplyPositionAlignedDiff →
///      ApplyRowToExistingModel → PropertyChanged("IsInfoRow") → UpdateAllComponents).
///
/// The AutomationId-based assertions check the Grid.Children state indirectly:
///   - "TimetableRow.N.StationName" is present &amp; user-visible before the transition.
///   - After the transition: StationName is gone; "TimetableRow.N.InfoRow" is visible.
/// </summary>
[TestFixture]
[Infrastructure.RetryAllTests(2)]
public class IsInfoRowTransitionTests : Infrastructure.BaseUITest
{
	private StartHomePageObject _startHomePage = null!;

	protected override bool ShareSessionAcrossTestsInFixture => true;

	[SetUp]
	public override void SetUp()
	{
		base.SetUp();

		_startHomePage = new StartHomePageObject(Driver);

		if (!_startHomePage.PollDisplayed(AutomationIds.StartHome.Title, timeoutSeconds: 3))
		{
			new AppShellPage(Driver).NavigateToHome();
			_startHomePage = new StartHomePageObject(Driver);
		}
		_startHomePage.ClearLoaderForTesting();
		_startHomePage.AcceptPrivacyPolicyIfNeeded();
	}

	/// <summary>
	/// Core regression: after the IsInfoRow seam fires, the station-name label
	/// must disappear and the info-row label must appear.
	///
	/// Before the fix: UpdateAllComponents() in the IsInfoRow=true branch only
	/// called UpdateInfoRow() — station-name (and other non-InfoRow components)
	/// stayed in the Grid. This test would find StationName still visible → fail.
	///
	/// After the fix: RemoveNonInfoRowComponents() is called first, so StationName
	/// is removed from the Grid before UpdateInfoRow() adds the InfoRow label.
	/// </summary>
	[Test]
	public void InfoRowTransition_FalseToTrue_RemovesStationNameShowsInfoLabel()
	{
		// --- Setup: load sample data and navigate to DTAC timetable. ---
		Assert.That(_startHomePage.IsDisplayed(), Is.True);
		_startHomePage.LoadSample();
		_startHomePage.WaitForElement(AutomationIds.StartHome.WorkGroupList);

		var dtac = _startHomePage.AutoOpenForTesting();
		dtac.SwitchToTimetableTab();

		// Row 0 in the default sample train is "駅１" (IsInfoRow=false).
		// The UI_TEST build sets AutomationId = "TimetableRow.0.StationName" on it.
		string stationId = string.Format(AutomationIds.DTAC.TimetableRowStationNamePattern, 0);
		string infoRowId = string.Format(AutomationIds.DTAC.TimetableRowInfoRowPattern, 0);

		// --- Pre-condition: station-name label is visible before the transition. ---
		Assert.That(
			IsElementUserVisible(stationId),
			Is.True,
			"Row 0 StationNameLabel must be visible before the IsInfoRow transition.");

		// --- Action: trigger the seam that simulates the WebSocket IsInfoRow edit. ---
		var seamButton = dtac.WaitForElement(AutomationIds.DTAC.TestSeedIsInfoRowTransitionButton);
		seamButton.Click();
		Thread.Sleep(500); // allow the PropertyChanged cascade to settle

		// --- Post-condition (FAILS before fix): station-name must be gone. ---
		Assert.That(
			IsElementUserVisible(stationId),
			Is.False,
			"Row 0 StationNameLabel must NOT be visible after IsInfoRow changed to true " +
			"(non-InfoRow components must be removed from the Grid).");

		// --- Post-condition: info-row label must be visible. ---
		Assert.That(
			IsElementUserVisible(infoRowId),
			Is.True,
			"Row 0 InfoRowLabel must be visible after IsInfoRow changed to true.");
	}

	/// <summary>
	/// Returns true when the element is findable AND laid out (non-zero size).
	///
	/// Size-only (not <c>Displayed</c>) is the cross-platform-reliable signal because
	/// <c>Displayed</c> reports false in two unrelated, expected situations here:
	///   1. Mac Catalyst surfaces unparented elements with an AutomationId in the
	///      accessibility tree but reports them as 0×0 — size catches this.
	///   2. On iPhone the timetable Grid (width ≈ 740) is wider than the screen
	///      (≈ 390 pt), so XCUITest cascades <c>visible=false</c> from the off-screen
	///      parent down to every child. The child's frame is still non-zero in that
	///      state, so a size check correctly treats it as present while
	///      <c>Displayed</c> would wrongly report it absent.
	/// When the component is actually removed from the Grid, the element either
	/// leaves the tree (iOS/Android) or reports 0×0 (macOS) — both make this false.
	/// </summary>
	private bool IsElementUserVisible(string automationId, double timeoutSeconds = 3)
	{
		var prevWait = TimeSpan.FromSeconds(10);
		var deadline = DateTime.UtcNow.AddSeconds(timeoutSeconds);
		var locator = AutomationIdLocator(automationId);
		try
		{
			Driver.Manage().Timeouts().ImplicitWait = TimeSpan.Zero;
			while (DateTime.UtcNow < deadline)
			{
				var elements = Driver.FindElements(locator);
				if (elements.Count > 0)
				{
					try
					{
						var el = elements[0];
						if (el.Size.Width > 0 && el.Size.Height > 0)
							return true;
					}
					catch { }
				}
				Thread.Sleep(100);
			}
			return false;
		}
		finally
		{
			Driver.Manage().Timeouts().ImplicitWait = prevWait;
		}
	}

	private By AutomationIdLocator(string automationId)
		=> IsAndroid ? By.Id(automationId) : MobileBy.AccessibilityId(automationId);
}

File: TRViS.UITests.Apple/Sources/Shared/IsInfoRowTransitionTests.swift
// IsInfoRowTransitionTests.swift
// XCUITest port of TRViS.UITests/Tests/IsInfoRowTransitionTests.cs (Phase 2C).
//
// Regression for "non-InfoRow components (station name, drive time, etc.)
// remain visible after a WebSocket soft-update changes IsInfoRow from false
// to true on a timetable row."
//
// The C# fixture is [Platform(Exclude = "Linux")] — the Android UIAutomator2
// limitation (Linux NUnit host) does NOT apply here. Apple is fully supported.
//
// Per-test cold launch is used (BaseUITestCase.setUpWithError launches fresh).

import XCTest

final class IsInfoRowTransitionTests: BaseUITestCase {

    private var startHome: StartHomePageObject!

    override func setUpWithError() throws {
        try super.setUpWithError()
        startHome = StartHomePageObject(app: app, base: self)
        XCTAssertTrue(
            startHome.isDisplayed(),
            "StartHomePage should be displayed after launch (IsInfoRowTransitionTests setUp)."
        )
        startHome.acceptPrivacyPolicyIfNeeded()
        XCTAssertTrue(
            startHome.isDisplayed(),
            "StartHomePage should be displayed after accepting privacy."
        )
    }

    // MARK: — Test: InfoRowTransition_FalseToTrue_RemovesStationNameShowsInfoLabel

    /// Mirrors C# IsInfoRowTransitionTests.InfoRowTransition_FalseToTrue_RemovesStationNameShowsInfoLabel.
    ///
    /// Core regression: after the IsInfoRow seam fires, the station-name label
    /// must disappear and the info-row label must appear.
    ///
    /// Visibility is determined by frame size (width>0 && height>0) rather than
    /// isHittable. On iPhone the timetable Grid (~740 pt wide) is wider than the
    /// screen (~390 pt), so XCUITest cascades isHittable=false from the off-screen
    /// parent. Frame size is non-zero in that state — matching the C# heuristic.
    func testInfoRowTransition_FalseToTrue_RemovesStationNameShowsInfoLabel() throws {
        XCTAssertTrue(startHome.isDisplayed())
        startHome.loadSample()
        _ = startHome.waitForWorkGroupList(timeout: 30)

        let dtac = startHome.autoOpenForTesting()
        dtac.switchToTimetableTab()

        // Row 0 in the default sample train is "駅１" (IsInfoRow=false).
        let stationId = String(format: AutomationIds.DTAC.timetableRowStationNamePattern, 0)
        let infoRowId = String(format: AutomationIds.DTAC.timetableRowInfoRowPattern, 0)

        // Pre-condition: station-name label is visible before the transition.
        XCTAssertTrue(
            dtac.isElementUserVisible(automationId: stationId, timeout: 3),
            "Row 0 StationNameLabel must be visible before the IsInfoRow transition."
        )

        // Action: trigger the seam that simulates the WebSocket IsInfoRow edit.
        dtac.tapSeedIsInfoRowTransition()
        Thread.sleep(forTimeInterval: 0.5) // allow the PropertyChanged cascade to settle

        // Post-condition: station-name must be gone.
        // isElementUserVisible polls until timeout; if it returns false, the element
        // either left the tree or collapsed to 0×0 — both are the correct outcome.
        XCTAssertFalse(
            dtac.isElementUserVisible(automationId: stationId, timeout: 3),
            "Row 0 StationNameLabel must NOT be visible after IsInfoRow changed to true " +
            "(non-InfoRow components must be removed from the Grid)."
        )

        // Post-condition: info-row label must be visible.
        XCTAssertTrue(
            dtac.isElementUserVisible(automationId: infoRowId, timeout: 3),
            "Row 0 InfoRowLabel must be visible after IsInfoRow changed to true."
        )
    }
}
